### PR TITLE
Fix #134, #135. Add ability to create SQL Server and SQL Database fro…

### LINKED ISSUE #134: [Azure Explorer] Add ability to create SQL Server via Azure Explorer
We should provide an ability to create a SQL Server through context menu for "Databases" node in Azure Explorer.

### LINKED ISSUE #135: [Azure Explorer] Add ability to create SQL Database via Azure Explorer
We should provide an ability to create a SQL Database through context menu for "Databases" > "SQL Server" node in Azure Explorer.

### PR DESCRIPTION
**Pull request by sdubov**
…m Azure Explorer

- Fix #134. Add Create SQL Server dialogs available from Azure Explorer from Databases node
- Fix #135. Add Create SQL Database dialog available from Azure Explorer from SQL Server node
- Fix issues #133
- Add custom components for Subscription, Resource Group, App Service Plan, SQL Server and helpers
- Refactor validation logic - move all common logic into a separate validators and add wrappers for Run configuraiton and dialogs validation
- Move AzureDatabaseMvpModel into plugin Utils project and refactor related logic
- Generalize Azure Notification constants available in Azure Event Log

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/component/AzureSqlServerSelector.kt [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        add(rdoCreateSqlServer)
+        pnlSqlServerName.add(txtSqlServerName, "growx")
+        pnlSqlServerName.add(lblSqlServerNameSuffix)
+        add(pnlSqlServerName, "growx")
+        add(lblCreateLocationName, "gapbefore 3")
+        add(cbLocation, "growx")
+        add(lblCreateAdminLoginName, "gapbefore 3")
+        add(txtAdminLogin, "growx")
+        add(lblCreateAdminPasswordName, "gapbefore 3")
+        add(passNewAdminPasswordValue, "growx")
```
Comment (maartenba) [inline PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/component/AzureSqlServerSelector.kt:90]: [Enhancement] Should this also have a button next to it - "Generate", which generates a password + copies to clipboard?

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/component/AzureSubscriptionsSelector.kt [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+ */
+
+package com.microsoft.intellij.component
+
+import com.intellij.openapi.ui.ComboBox
+import com.intellij.openapi.ui.ValidationInfo
+import com.intellij.ui.components.Label
+import com.intellij.util.ui.UIUtil
+import com.microsoft.azure.management.resources.Subscription
+import com.microsoft.intellij.component.extention.createDefaultRenderer
```
Comment (maartenba) [inline PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/component/AzureSubscriptionsSelector.kt]: Spelling: `extension` (more occurrences throughout this PR)

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/forms/sqldatabase/CreateSqlDatabaseOnServerDialog.kt [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+        private const val TITLE_RESOURCE_GROUP = "Resource Group"
+        private const val TITLE_SQL_DATABASE_SETTINGS = "Database Settings"
+
+        private const val SQL_DATABASE_CREATING_MESSAGE = "Creating '%s' database..."
+        private const val SQL_DATABASE_CREATE_SUCCESSFUL = "SQL Database is created, id: '%s'"
+
+        private const val AZURE_SQL_DATABASE_HELP_URL = "https://azure.microsoft.com/en-us/services/sql-database/"
+        private const val AZURE_SQL_DATABASE_PRICING_URI = "https://azure.microsoft.com/en-us/pricing/details/sql-database/"
+
+        private const val SQL_DATABASE_CREATE_TIMEOUT_MS = 60_000L
```
Comment (maartenba) [inline PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/forms/sqldatabase/CreateSqlDatabaseOnServerDialog.kt]: Should this be > 1 minute? Have seen database creation take longer, perhaps make it 2 minutes?

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/forms/sqlserver/CreateSqlServerDialog.kt [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+        private const val DIALOG_MIN_WIDTH = 300
+
+        private const val TITLE_RESOURCE_GROUP = "Resource Group"
+        private const val TITLE_SQL_SERVER_SETTINGS = "Server Settings"
+
+        private const val SQL_SERVER_CREATING_MESSAGE = "Creating '%s' SQL Server..."
+        private const val SQL_SERVER_CREATE_SUCCESSFUL = "SQL Server is created, id: '%s'"
+        private const val AZURE_SQL_SERVER_HELP_URL = "https://azure.microsoft.com/en-us/services/sql-database/"
+
+        private const val SQL_SERVER_CREATE_TIMEOUT_MS = 60_000L
```
Comment (maartenba) [inline PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/forms/sqlserver/CreateSqlServerDialog.kt]: 2 minutes?

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/runner/webapp/webappconfig/ui/RiderWebAppSettingPanel.kt [hunk 5]
```diff
@@ -27,6 +27,7 @@ import com.intellij.ide.BrowserUtil
 import com.intellij.ide.util.PropertiesComponent
 import com.intellij.openapi.actionSystem.ActionToolbarPosition
 import com.intellij.openapi.actionSystem.AnActionEvent
+import com.intellij.openapi.application.ApplicationManager
```
Comment (maartenba) [inline PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/runner/webapp/webappconfig/ui/RiderWebAppSettingPanel.kt:30]: Line 1573: `return targetFramework?.currentTargetFrameworkId?.valueOrNull`

Does not compile on my box. Has to be:

`return targetFramework?.currentTargetFrameworkId?.valueOrNull?.id`

**Review by maartenba — CHANGES_REQUESTED** [review]
Reviewed code, couple of comments.

**Comment by maartenba** [comment]
The word/package `extention` should be `extension`.

**Comment by maartenba** [comment]
`README.md` needs a bit of an update (what was added)

**Comment by maartenba** [comment]
Ran the plugin and experimented a bit, found the following issues:

* Create DB with Free tier, got a paid one. Not sure if this is in our code or something wrong in the Azure API's. (in fact I do not thing there is a free tier anymore)

* Delete server resulted in a hang of Rider. Will send logs privately.

* When creating server, got: (note the server was created after a while)

```
java.lang.RuntimeException: com.microsoft.azuretools.adauth.AuthException: AD Auth token endpoint returned HTTP status code 400. Error info: {"error":"invalid_request","error_description":"AADSTS90002: Tenant null not found. This may happen if there are no active subscriptions for the tenant. Check with your subscription administrator.\r\nTrace ID: a8177905-8669-4ecf-b926-c2f51f024700\r\nCorrelation ID: bba8d5be-13b2-4090-9c12-0de182ed98b7\r\nTimestamp: 2018-12-24 07:50:40Z","error_codes":[90002],"timestamp":"2018-12-24 07:50:40Z","trace_id":"a8177905-8669-4ecf-b926-c2f51f024700","correlation_id":"bba8d5be-13b2-4090-9c12-0de182ed98b7"}
	at rx.exceptions.Exceptions.propagate(Exceptions.java:57)
	at rx.observables.BlockingObservable.blockForSingle(BlockingObservable.java:463)
	at rx.observables.BlockingObservable.single(BlockingObservable.java:340)
	at com.microsoft.azure.management.sql.implementation.ServersInner.list(ServersInner.java:208)
	at com.microsoft.azure.management.resources.fluentcore.arm.collection.implementation.TopLevelModifiableResourcesImpl.list(TopLevelModifiableResourcesImpl.java:118)
	at com.microsoft.azuretools.core.mvp.model.database.AzureSqlServerMvpModel.getSqlServersBySubscriptionId(AzureSqlServerMvpModel.kt:141)
	at com.microsoft.azuretools.core.mvp.model.database.AzureSqlServerMvpModel.listSqlServersBySubscriptionId(AzureSqlServerMvpModel.kt:61)
	at com.microsoft.azuretools.core.mvp.model.database.AzureSqlServerMvpModel.refreshSubscriptionToSqlServerMap(AzureSqlServerMvpModel.kt:45)
	at com.microsoft.azuretools.core.mvp.model.database.AzureSqlDatabaseMvpModel.refreshSqlServerToSqlDatabaseMap(AzureSqlDatabaseMvpModel.kt:47)
	at com.microsoft.intellij.forms.sqlserver.CreateSqlServerDialog$updateAzureModelInBackground$1.run(CreateSqlServerDialog.kt:299)
	at com.intellij.openapi.progress.impl.CoreProgressManager$TaskRunnable.run(CoreProgressManager.java:727)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:164)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:582)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:532)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:86)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:151)
	at com.intellij.openapi.progress.impl.CoreProgressManager$4.run(CoreProgressManager.java:403)
	at com.intellij.openapi.application.impl.ApplicationImpl$1.run(ApplicationImpl.java:315)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: com.microsoft.azuretools.adauth.AuthException: AD Auth token endpoint returned HTTP status code 400. Error info: {"error":"invalid_request","error_description":"AADSTS90002: Tenant null not found. This may happen if there are no active subscriptions for the tenant. Check with your subscription administrator.\r\nTrace ID: a8177905-8669-4ecf-b926-c2f51f024700\r\nCorrelation ID: bba8d5be-13b2-4090-9c12-0de182ed98b7\r\nTimestamp: 2018-12-24 07:50:40Z","error_codes":[90002],"timestamp":"2018-12-24 07:50:40Z","trace_id":"a8177905-8669-4ecf-b926-c2f51f024700","correlation_id":"bba8d5be-13b2-4090-9c12-0de182ed98b7"}
	at com.microsoft.azuretools.adauth.AuthContext.sendRequest(AuthContext.java:207)
	at com.microsoft.azuretools.adauth.AuthContext.access$100(AuthContext.java:41)
	at com.microsoft.azuretools.adauth.AuthContext$1.getTokenWithRefreshToken(AuthContext.java:192)
	at com.microsoft.azuretools.adauth.CacheDriver.acquireTokenFromMrrt(CacheDriver.java:117)
	at com.microsoft.azuretools.adauth.CacheDriver.refreshEntryIfNecessary(CacheDriver.java:109)
	at com.microsoft.azuretools.adauth.CacheDriver.find(CacheDriver.java:55)
	at com.microsoft.azuretools.adauth.AuthContext.acquireTokenFromCache(AuthContext.java:161)
	at com.microsoft.azuretools.adauth.AuthContext.acquireToken(AuthContext.java:122)
	at com.microsoft.azuretools.authmanage.AdAuthManager.getAccessToken(AdAuthManager.java:92)
	at com.microsoft.azuretools.authmanage.RefreshableTokenCredentials.getToken(RefreshableTokenCredentials.java:65)
	at com.microsoft.azure.credentials.AzureTokenCredentials.getToken(AzureTokenCredentials.java:74)
	at com.microsoft.azure.credentials.AzureTokenCredentialsInterceptor.intercept(AzureTokenCredentialsInterceptor.java:36)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at com.microsoft.azure.management.resources.fluentcore.utils.ResourceManagerThrottlingInterceptor.intercept(ResourceManagerThrottlingInterceptor.java:51)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at com.microsoft.azure.management.resources.fluentcore.utils.ProviderRegistrationInterceptor.intercept(ProviderRegistrationInterceptor.java:43)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at com.microsoft.azuretools.telemetry.TelemetryInterceptor.intercept(TelemetryInterceptor.java:46)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at com.microsoft.rest.interceptors.BaseUrlHandler.intercept(BaseUrlHandler.java:43)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at com.microsoft.rest.interceptors.RequestIdHeaderInterceptor.intercept(RequestIdHeaderInterceptor.java:29)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:200)
	at okhttp3.RealCall.execute(RealCall.java:77)
	at retrofit2.OkHttpCall.execute(OkHttpCall.java:180)
	at retrofit2.adapter.rxjava.CallExecuteOnSubscribe.call(CallExecuteOnSubscribe.java:40)
	at retrofit2.adapter.rxjava.CallExecuteOnSubscribe.call(CallExecuteOnSubscribe.java:24)
	at rx.Observable.unsafeSubscribe(Observable.java:10327)
	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:48)
	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:33)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
	at rx.Observable.subscribe(Observable.java:10423)
	at rx.Observable.subscribe(Observable.java:10390)
	at rx.observables.BlockingObservable.blockForSingle(BlockingObservable.java:443)
	... 21 more
Caused by: java.io.IOException: AD Auth token endpoint returned HTTP status code 400. Error info: {"error":"invalid_request","error_description":"AADSTS90002: Tenant null not found. This may happen if there are no active subscriptions for the tenant. Check with your subscription administrator.\r\nTrace ID: a8177905-8669-4ecf-b926-c2f51f024700\r\nCorrelation ID: bba8d5be-13b2-4090-9c12-0de182ed98b7\r\nTimestamp: 2018-12-24 07:50:40Z","error_codes":[90002],"timestamp":"2018-12-24 07:50:40Z","trace_id":"a8177905-8669-4ecf-b926-c2f51f024700","correlation_id":"bba8d5be-13b2-4090-9c12-0de182ed98b7"}
	at com.microsoft.azuretools.adauth.HttpHelper.sendPostRequestAndDeserializeJsonResponse(HttpHelper.java:92)
	at com.microsoft.azuretools.adauth.AuthContext.sendRequest(AuthContext.java:203)
	... 64 more
```

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/component/AzureSqlServerSelector.kt (hunk 1 above)
Comment (sdubov) [reply PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/component/AzureSqlServerSelector.kt:90]: There are several things to think about before adding this functionality. So, create a separate ticket fo that - https://github.com/JetBrains/azure-tools-for-intellij/issues/138

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/component/AzureSubscriptionsSelector.kt (hunk 2 above)
Comment (sdubov) [reply PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/component/AzureSubscriptionsSelector.kt]: Fixed

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/forms/sqldatabase/CreateSqlDatabaseOnServerDialog.kt (hunk 3 above)
Comment (sdubov) [reply PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/forms/sqldatabase/CreateSqlDatabaseOnServerDialog.kt]: Set to 2 mins

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/runner/webapp/webappconfig/ui/RiderWebAppSettingPanel.kt (hunk 5 above)
Comment (sdubov) [reply PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/runner/webapp/webappconfig/ui/RiderWebAppSettingPanel.kt:30]: Fixed. Rider SDK has been changed.

**Comment by maartenba** [comment]
Looks good @sdubov ! Did you manage to reproduce the hang and Exception as well?

**Comment by sdubov** [comment]
@maartenba, the hang is caused by running SQL Server delete on UI thread (twice!). Fixed in a recent patch. The exception is an azure core issue. I faced it several times. We have a related DEXP for this issue as far as I remember - https://youtrack.jetbrains.com/issue/DEXP-436593. I'll take a look on it in 191.

**Review by maartenba — APPROVED** [review]

**Comment by maartenba** [comment]
Can be merged I think?

**Comment by sdubov** [comment]
@maartenba, there are some fixes I took from this PR to 183.1 branch. I would first merge 183.1 into 191, then clean code in this branch and merge.

**Pull request merged**